File: ios/UGames/WebView/InjectedScripts.swift
import Foundation

struct InjectedScripts {
    let honestPath: String
    let pwaModeJs: String
    let pwaModeCss: String
    let sdkStub: String

    var mainFrameScript: String {
        let cssLiteral = jsString(pwaModeCss)
        return """
        window.__yga_pwa_css_payload__=\(cssLiteral);
        \(honestPath);
        \(pwaModeJs)
        """
    }

    static func load() -> InjectedScripts {
        InjectedScripts(
            honestPath: readResource("honest-path", ext: "js"),
            pwaModeJs: readResource("pwa-mode", ext: "js"),
            pwaModeCss: readResource("pwa-mode", ext: "css"),
            sdkStub: readResource("ya-sdk-stub", ext: "js")
        )
    }

    private static func readResource(_ name: String, ext: String) -> String {
        guard let url = Bundle.main.url(forResource: name, withExtension: ext) else {
            assertionFailure("Missing bundle resource \(name).\(ext)")
            return ""
        }
        return (try? String(contentsOf: url, encoding: .utf8)) ?? ""
    }

    private func jsString(_ value: String) -> String {
        var out = "\""
        for char in value.unicodeScalars {
            switch char.value {
            case 0x5C: out.append("\\\\")
            case 0x22: out.append("\\\"")
            case 0x0A: out.append("\\n")
            case 0x0D: out.append("\\r")
            case 0x09: out.append("\\t")
            case 0x2028: out.append("\\u2028")
            case 0x2029: out.append("\\u2029")
            default: out.append(Character(char))
            }
        }
        out.append("\"")
        return out
    }
}

File: android/app/src/main/kotlin/games/yandex/wrap/webview/InjectedScripts.kt
package games.yandex.wrap.webview

import android.content.Context

class InjectedScripts(
    val honestPath: String,
    val pwaModeJs: String,
    val pwaModeCss: String,
    val sdkStub: String,
) {
    val mainFrameScript: String by lazy {
        buildString {
            append(honestPath)
            append(";\n")
            append("(function(){var s=document.createElement('style');s.textContent=")
            append(jsString(pwaModeCss))
            append(";document.documentElement.appendChild(s);})();\n")
            append(pwaModeJs)
        }
    }

    companion object {
        fun load(context: Context): InjectedScripts = InjectedScripts(
            honestPath = readAsset(context, "honest-path.js"),
            pwaModeJs = readAsset(context, "pwa-mode.js"),
            pwaModeCss = readAsset(context, "pwa-mode.css"),
            sdkStub = readAsset(context, "ya-sdk-stub.js"),
        )

        private fun readAsset(context: Context, name: String): String =
            context.assets.open(name).bufferedReader(Charsets.UTF_8).use { it.readText() }

        private fun jsString(value: String): String {
            val sb = StringBuilder("\"")
            for (c in value) {
                when (c.code) {
                    0x5C -> sb.append("\\\\")
                    0x22 -> sb.append("\\\"")
                    0x0A -> sb.append("\\n")
                    0x0D -> sb.append("\\r")
                    0x09 -> sb.append("\\t")
                    0x2028 -> sb.append("\\u2028")
                    0x2029 -> sb.append("\\u2029")
                    else -> sb.append(c)
                }
            }
            sb.append("\"")
            return sb.toString()
        }
    }
}
